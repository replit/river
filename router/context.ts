import { Connection, Session } from '../transport/session';

/**
 * ServiceContext exist for the purpose of declaration merging
 * to extend the context with additional properties.
 *
 * For example:
 *
 * ```ts
 * declare module '@replit/river' {
 *   interface ServiceContext {
 *     db: Database;
 *   }
 * }
 *
 * createServer(someTransport, myServices, { extendedContext: { db: myDb } });
 * ```
 *
 * Once you do this, your {@link ProcedureHandlerContext} will have `db` property on it.
 */
/* eslint-disable-next-line @typescript-eslint/no-empty-interface */
export interface ServiceContext {}

/**
 * The parsed metadata schema for a service. This is the
 * return value of the {@link ServerHandshakeOptions.validate}
 * if the handshake extension is used.
 *
 * You should use declaration merging to extend this interface
 * with the sanitized metadata.
 *
 * ```ts
 * declare module '@replit/river' {
 *   interface ParsedMetadata {
 *     userId: number;
 *   }
 * }
 * ```
 */
/* eslint-disable-next-line @typescript-eslint/no-empty-interface */
export interface ParsedMetadata {}

/**
 * This is passed to every procedure handler and contains various context-level
 * information and utilities. This may be extended, see {@link ServiceContext}
 */
export type ProcedureHandlerContext<State> = ServiceContext & {
  /**
   * State for this service as defined by the service definition.
   */
  state: State;
  /**
   * Metadata parsed on the server. See {@link ParsedMetadata}
   */
  metadata: ParsedMetadata;
  /**
   * The session for this stream. Can be used to identify the request
   * sender via `session.from`, however, we recommend using the
   * handshake metadata with some auth mechanism to identify the
   * the request sender.
   */
  session: Session<Connection>; // TODO: only expose a subset interface of session
  /**
   * An AbortController for this stream. This is used to abort the stream from the
   * handler and notify the client that the stream was aborted.
   *
   * Important to note that this controller is owned by the handler, if you
   * want to listen to aborts coming from the client, you should use the
   * {@link clientAbortSignal}.
   *
   * Aborts are not the same as closing streams gracefully, please refer to
   * the river documentation to understand the difference between the two concepts.
   */
  abortController: AbortController;
  /**
   * You can listen to clientAbortSignal this to check if the client aborted the request,
   * or if the request was aborted due to an unexpected disconnect from the calling
   * session.
   *
   * If the procedure has a read stream (e.g. upload or stream), the procedure will
   * notified of aborts as part of the stream, but you may still want to use
   * this signal as it is triggered immediately after an abort comes in,
   * in readStreams some data may be buffered before the abort result shows up.
   *
   * Important to note that this signal is owned by the client, you have a separate
   * signal inside {@link abortController} for aborts triggered within the handler.
   *
   * Aborts are not the same as closing streams gracefully, please refer to
   * the river documentation to understand the difference between the two concepts.
   */
  clientAbortSignal: AbortSignal;
  /**
   * Lets you add a function that will run when the stream is cleaned up due to
   * it ending. If a cleanup is added after the stream has ended, it will
   * run immediately.
   */
  addCleanup: (cleanup: () => void) => void;
};
