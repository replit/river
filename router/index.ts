export type {
  Service,
  ServiceConfiguration,
  ProcHandler,
  ProcInit,
  ProcInput,
  ProcOutput,
  ProcErrors,
  ProcType,
} from './services';
export {
  ServiceSchema,
  serializeSchema,
  SerializedServerSchema,
} from './services';
export type {
  ValidProcType,
  PayloadType,
  ProcedureMap,
  RpcProcedure as RPCProcedure,
  UploadProcedure,
  SubscriptionProcedure,
  StreamProcedure,
  ProcedureErrorSchemaType,
} from './procedures';
export {
  Procedure,
  UNCAUGHT_ERROR_CODE,
  UNEXPECTED_DISCONNECT_CODE,
  INVALID_REQUEST_CODE,
  OutputReaderErrorSchema,
  InputReaderErrorSchema,
} from './procedures';
export { createClient } from './client';
export type { Client } from './client';
export { createServer } from './server';
export type { Server } from './server';
export type {
  ParsedMetadata,
  ServiceContext,
  ServiceContextWithState,
  ServiceContextWithTransportInfo,
} from './context';
export { Ok, Err } from './result';
export type {
  Result,
  ErrResult,
  OkResult,
  ResultUnwrapOk,
  ResultUnwrapErr,
  Output,
} from './result';
export {
  createClientHandshakeOptions,
  createServerHandshakeOptions,
} from './handshake';
export { version as RIVER_VERSION } from '../package.json';
