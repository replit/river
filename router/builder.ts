import { TObject, Static, Type, TUnion } from '@sinclair/typebox';
import type { Pushable } from 'it-pushable';
import { ServiceContextWithTransportInfo } from './context';
import { Result, RiverError, RiverUncaughtSchema } from './result';

/**
 * The valid {@link Procedure} types. The `stream` and `upload` types can optionally have a
 * different type for the very first initialization message. The suffixless types correspond to
 * gRPC's four combinations of stream / non-stream in each direction.
 */
export type ValidProcType =
  // Single message in both directions (1:1).
  | 'rpc'
  // Client-stream (potentially preceded by an initialization message), single message from server (n:1).
  | 'upload'
  // Single message from client, stream from server (1:n).
  | 'subscription'
  // Bidirectional stream (potentially preceded by an initialization message) (n:n).
  | 'stream';

/**
 * A generic procedure listing where the keys are the names of the procedures
 * and the values are the {@link Procedure} definitions. This is not meant to
 * be constructed directly, use the {@link ServiceBuilder} class instead.
 */
export type ProcListing = Record<string, AnyProcedure>;

/**
 * Represents a service with a name, state, and procedures.
 * This is not meant to be constructed directly, use the {@link ServiceBuilder} class instead.
 * @template Name The type of the service name.
 * @template State The type of the service state.
 * @template Procs The type of the service procedures.
 */
export interface Service<
  Name extends string,
  State extends object,
  // nested record (service listing contains services which have proc listings)
  // this means we lose type specificity on our procedures here so we maintain it by using
  // any on the default type
  Procs extends ProcListing,
> {
  name: Name;
  state: State;
  procedures: Procs;
}
export type AnyService = Service<string, object, ProcListing>;

/**
 * Serializes a service object into its corresponding JSON Schema Draft 7 type.
 * @param {AnyService} s - The service object to serialize.
 * @returns A plain object representing the serialized service.
 */
export function serializeService(s: AnyService): object {
  return {
    name: s.name,
    state: s.state,
    procedures: Object.fromEntries(
      Object.entries<AnyProcedure>(s.procedures).map(([procName, procDef]) => [
        procName,
        {
          input: Type.Strict(procDef.input),
          output: Type.Strict(procDef.output),
          // Only add the `errors` field if it is non-never.
          ...('errors' in procDef
            ? {
                errors: Type.Strict(procDef.errors),
              }
            : {}),
          type: procDef.type,
          // Only add the `init` field if the type declares it.
          ...('init' in procDef
            ? {
                init: Type.Strict(procDef.init),
              }
            : {}),
        },
      ]),
    ),
  };
}

/**
 * Helper to get the type definition for a specific handler of a procedure in a service.
 * @template S - The service.
 * @template ProcName - The name of the procedure.
 */
export type ProcHandler<
  S extends AnyService,
  ProcName extends keyof S['procedures'],
> = S['procedures'][ProcName]['handler'];

/**
 * Helper to get whether the type definition for the procedure contains an init type.
 * @template S - The service.
 * @template ProcName - The name of the procedure.
 */
export type ProcHasInit<
  S extends AnyService,
  ProcName extends keyof S['procedures'],
> = S['procedures'][ProcName] extends { init: TObject } ? true : false;

/**
 * Helper to get the type definition for the procedure init type of a service.
 * @template S - The service.
 * @template ProcName - The name of the procedure.
 */
export type ProcInit<
  S extends AnyService,
  ProcName extends keyof S['procedures'],
> = S['procedures'][ProcName] extends { init: TObject }
  ? S['procedures'][ProcName]['init']
  : never;

/**
 * Helper to get the type definition for the procedure input of a service.
 * @template S - The service.
 * @template ProcName - The name of the procedure.
 */
export type ProcInput<
  S extends AnyService,
  ProcName extends keyof S['procedures'],
> = S['procedures'][ProcName]['input'];

/**
 * Helper to get the type definition for the procedure output of a service.
 * @template S - The service.
 * @template ProcName - The name of the procedure.
 */
export type ProcOutput<
  S extends AnyService,
  ProcName extends keyof S['procedures'],
> = S['procedures'][ProcName]['output'];

/**
 * Helper to get the type definition for the procedure errors of a service.
 * @template S - The service.
 * @template ProcName - The name of the procedure.
 */
export type ProcErrors<
  S extends AnyService,
  ProcName extends keyof S['procedures'],
> = TUnion<[S['procedures'][ProcName]['errors'], typeof RiverUncaughtSchema]>;

/**
 * Helper to get the type of procedure in a service.
 * @template S - The service.
 * @template ProcName - The name of the procedure.
 */
export type ProcType<
  S extends AnyService,
  ProcName extends keyof S['procedures'],
> = S['procedures'][ProcName]['type'];

export type PayloadType = TObject | TUnion<Array<TObject>>;

/**
 * Defines a Procedure type that can be either an RPC or a stream procedure.
 * @template State - The TypeBox schema of the state object.
 * @template Ty - The type of the procedure.
 * @template I - The TypeBox schema of the input object.
 * @template O - The TypeBox schema of the output object.
 * @template Init - The TypeBox schema of the input initialization object.
 */
export type Procedure<
  State,
  Ty extends ValidProcType,
  I extends PayloadType,
  O extends PayloadType,
  E extends RiverError,
  Init extends PayloadType | null = null,
> = Ty extends 'rpc'
  ? Init extends null
    ? {
        input: I;
        output: O;
        errors: E;
        handler: (
          context: ServiceContextWithTransportInfo<State>,
          input: Static<I>,
        ) => Promise<Result<Static<O>, Static<E>>>;
        type: Ty;
      }
    : never
  : Ty extends 'upload'
  ? Init extends PayloadType
    ? {
        init: Init;
        input: I;
        output: O;
        errors: E;
        handler: (
          context: ServiceContextWithTransportInfo<State>,
          init: Static<Init>,
          input: AsyncIterableIterator<Static<I>>,
        ) => Promise<Result<Static<O>, Static<E>>>;
        type: Ty;
      }
    : {
        input: I;
        output: O;
        errors: E;
        handler: (
          context: ServiceContextWithTransportInfo<State>,
          input: AsyncIterableIterator<Static<I>>,
        ) => Promise<Result<Static<O>, Static<E>>>;
        type: Ty;
      }
  : Ty extends 'subscription'
  ? Init extends null
    ? {
        input: I;
        output: O;
        errors: E;
        handler: (
          context: ServiceContextWithTransportInfo<State>,
          input: Static<I>,
          output: Pushable<Result<Static<O>, Static<E>>>,
        ) => Promise<(() => void) | void>;
        type: Ty;
      }
    : never
  : Ty extends 'stream'
  ? Init extends PayloadType
    ? {
        init: Init;
        input: I;
        output: O;
        errors: E;
        handler: (
          context: ServiceContextWithTransportInfo<State>,
          init: Static<Init>,
          input: AsyncIterableIterator<Static<I>>,
          output: Pushable<Result<Static<O>, Static<E>>>,
        ) => Promise<void>;
        type: Ty;
      }
    : {
        input: I;
        output: O;
        errors: E;
        handler: (
          context: ServiceContextWithTransportInfo<State>,
          input: AsyncIterableIterator<Static<I>>,
          output: Pushable<Result<Static<O>, Static<E>>>,
        ) => Promise<void>;
        type: Ty;
      }
  : never;
export type AnyProcedure = Procedure<
  object,
  ValidProcType,
  PayloadType,
  PayloadType,
  RiverError,
  PayloadType | null
>;

/**
 * A builder class for creating River Services.
 * You must call the finalize method to get the finalized schema for use in a service router.
 * @template T The type of the service.
 */
export class ServiceBuilder<T extends Service<string, object, ProcListing>> {
  private readonly schema: T;
  private constructor(schema: T) {
    this.schema = schema;
  }

  /**
   * Finalizes the schema for the service.
   */
  finalize() {
    return Object.freeze(this.schema);
  }

  /**
   * Sets the initial state for the service.
   * @template InitState The type of the initial state.
   * @param {InitState} state The initial state for the service.
   * @returns {ServiceBuilder<{ name: T['name']; state: InitState; procedures: T['procedures']; }>} A new ServiceBuilder instance with the updated schema.
   */
  initialState<InitState extends T['state']>(
    state: InitState,
  ): ServiceBuilder<{
    name: T['name'];
    state: InitState;
    procedures: T['procedures'];
  }> {
    return new ServiceBuilder({
      ...this.schema,
      state,
    });
  }

  /**
   * Defines a new procedure for the service.
   * @param {ProcName} procName The name of the procedure.
   * @param {Procedure<T['state'], Ty, I, O, E, Init>} procDef The definition of the procedure.
   * @returns {ServiceBuilder<{ name: T['name']; state: T['state']; procedures: T['procedures'] & { [k in ProcName]: Procedure<T['state'], Ty, I, O, E, Init>; }; }>} A new ServiceBuilder instance with the updated schema.
   */
  defineProcedure<
    ProcName extends string,
    Ty extends ValidProcType,
    I extends PayloadType,
    O extends PayloadType,
    E extends RiverError,
    Init extends PayloadType | null = null,
  >(
    procName: ProcName,
    procDef: Procedure<T['state'], Ty, I, O, E, Init>,
  ): ServiceBuilder<{
    name: T['name'];
    state: T['state'];
    procedures: T['procedures'] & {
      [k in ProcName]: Procedure<T['state'], Ty, I, O, E, Init>;
    };
  }> {
    type ProcListing = {
      [k in ProcName]: Procedure<T['state'], Ty, I, O, E, Init>;
    };
    const newProcedure = { [procName]: procDef } as ProcListing;
    const procedures = {
      ...this.schema.procedures,
      ...newProcedure,
    } as {
      [Key in keyof (T['procedures'] & ProcListing)]: (T['procedures'] &
        ProcListing)[Key];
    };
    return new ServiceBuilder({
      ...this.schema,
      procedures,
    });
  }

  /**
   * Creates a new instance of ServiceBuilder.
   * @param {Name} name The name of the service.
   * @returns {ServiceBuilder<{ name: Name; state: {}; procedures: {}; }>} A new instance of ServiceBuilder.
   */
  static create<Name extends string>(
    name: Name,
  ): ServiceBuilder<{
    name: Name;
    state: object;
    procedures: ProcListing;
  }> {
    return new ServiceBuilder({
      name,
      state: {},
      procedures: {},
    });
  }
}
