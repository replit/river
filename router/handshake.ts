import { Static, TSchema } from '@sinclair/typebox';
import { ParsedMetadata } from './context';
import { HandshakeErrorCustomHandlerFatalResponseCodes } from '../transport/message';

type ConstructHandshake<T extends TSchema> = () =>
  | Static<T>
  | Promise<Static<T>>;

type ValidateHandshake<T extends TSchema> = (
  metadata: Static<T>,
  previousParsedMetadata?: ParsedMetadata,
) =>
  | Static<typeof HandshakeErrorCustomHandlerFatalResponseCodes>
  | ParsedMetadata
  | Promise<
      | Static<typeof HandshakeErrorCustomHandlerFatalResponseCodes>
      | ParsedMetadata
    >;

export interface ClientHandshakeOptions<
  MetadataSchema extends TSchema = TSchema,
> {
  /**
   * Schema for the metadata that the client sends to the server
   * during the handshake.
   */
  schema: MetadataSchema;

  /**
   * Gets the {@link HandshakeRequestMetadata} to send to the server.
   */
  construct: ConstructHandshake<MetadataSchema>;
}

export interface ServerHandshakeOptions<
  MetadataSchema extends TSchema = TSchema,
> {
  /**
   * Schema for the metadata that the server receives from the client
   * during the handshake.
   */
  schema: MetadataSchema;

  /**
   * Parses the {@link HandshakeRequestMetadata} sent by the client, transforming
   * it into {@link ParsedHandshakeMetadata}.
   *
   * May return `false` if the client should be rejected.
   *
   * @param metadata - The metadata sent by the client.
   * @param session - The session that the client would be associated with.
   * @param isReconnect - Whether the client is reconnecting to the session,
   *                      or if this is a new session.
   */
  validate: ValidateHandshake<MetadataSchema>;
}

export function createClientHandshakeOptions<
  MetadataSchema extends TSchema = TSchema,
>(
  schema: MetadataSchema,
  construct: ConstructHandshake<MetadataSchema>,
): ClientHandshakeOptions {
  return { schema, construct };
}

export function createServerHandshakeOptions<
  MetadataSchema extends TSchema = TSchema,
>(
  schema: MetadataSchema,
  validate: ValidateHandshake<MetadataSchema>,
): ServerHandshakeOptions {
  return { schema, validate: validate as ValidateHandshake<TSchema> };
}
