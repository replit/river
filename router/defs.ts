import { AnyService } from './builder';

/**
 * Defines a type for a collection service definitions. Should be
 * build with the {@link buildServiceDefs} function.
 * @template T - An array of services.
 */
export type ServiceDefs<T extends Array<AnyService> = Array<AnyService>> = {
  [K in T[number]['name']]: Extract<T[number], { name: K }>;
};

/**
 * Builds service definitions based on an array of services.
 * @param services - The array of services.
 * @returns The service definitions.
 */
export function buildServiceDefs<T extends Array<AnyService>>(
  services: T,
): ServiceDefs<T> {
  // we use reduce for building objects from arrays (buildServiceDefs) which typescript cannot prove to be safe
  return services.reduce((acc, service) => {
    acc[service.name as keyof ServiceDefs<T>] = service as Extract<
      T[number],
      { name: T[number]['name'] }
    >;
    return acc;
    /* eslint-disable-next-line @typescript-eslint/prefer-reduce-type-parameter */
  }, {} as ServiceDefs<T>);
}
