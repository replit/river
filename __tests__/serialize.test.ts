import { expect, describe, test } from 'vitest';
import {
  BinaryFileServiceSchema,
  FallibleServiceSchema,
  TestServiceSchema,
} from './fixtures/services';
import { serializeSchema } from '../router';
import { Type } from '@sinclair/typebox';

describe('serialize server to jsonschema', () => {
  test('serialize entire service schema', () => {
    const schema = { test: TestServiceSchema };
    const handshakeSchema = Type.Object({
      token: Type.String(),
    });

    expect(serializeSchema(schema, handshakeSchema)).toMatchSnapshot();
  });
});

describe('serialize service to jsonschema', () => {
  test('serialize basic service', () => {
    expect(TestServiceSchema.serialize()).toMatchSnapshot();
  });

  test('serialize service with binary', () => {
    expect(BinaryFileServiceSchema.serialize()).toMatchSnapshot();
  });

  test('serialize service with errors', () => {
    expect(FallibleServiceSchema.serialize()).toMatchSnapshot();
  });

  test('serialize backwards compatible with v1', () => {
    expect(TestServiceSchema.serializeV1Compat()).toMatchSnapshot();
  });
});
