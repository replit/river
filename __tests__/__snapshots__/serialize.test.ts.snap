// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serialize server to jsonschema > serialize entire service schema 1`] = `
{
  "handshakeSchema": {
    "properties": {
      "token": {
        "type": "string",
      },
    },
    "required": [
      "token",
    ],
    "type": "object",
  },
  "services": {
    "test": {
      "procedures": {
        "add": {
          "errors": {
            "not": {},
          },
          "init": {
            "properties": {
              "n": {
                "type": "number",
              },
            },
            "required": [
              "n",
            ],
            "type": "object",
          },
          "output": {
            "properties": {
              "result": {
                "type": "number",
              },
            },
            "required": [
              "result",
            ],
            "type": "object",
          },
          "type": "rpc",
        },
        "array": {
          "errors": {
            "not": {},
          },
          "init": {
            "properties": {
              "n": {
                "type": "number",
              },
            },
            "required": [
              "n",
            ],
            "type": "object",
          },
          "output": {
            "items": {
              "type": "number",
            },
            "type": "array",
          },
          "type": "rpc",
        },
        "arrayStream": {
          "errors": {
            "not": {},
          },
          "init": {
            "properties": {},
            "type": "object",
          },
          "input": {
            "properties": {
              "n": {
                "type": "number",
              },
            },
            "required": [
              "n",
            ],
            "type": "object",
          },
          "output": {
            "items": {
              "type": "number",
            },
            "type": "array",
          },
          "type": "stream",
        },
        "echo": {
          "errors": {
            "not": {},
          },
          "init": {
            "properties": {},
            "type": "object",
          },
          "input": {
            "properties": {
              "ignore": {
                "type": "boolean",
              },
              "msg": {
                "type": "string",
              },
            },
            "required": [
              "msg",
              "ignore",
            ],
            "type": "object",
          },
          "output": {
            "properties": {
              "response": {
                "type": "string",
              },
            },
            "required": [
              "response",
            ],
            "type": "object",
          },
          "type": "stream",
        },
        "echoUnion": {
          "description": "Echos back whatever we sent",
          "errors": {
            "not": {},
          },
          "init": {
            "anyOf": [
              {
                "description": "A",
                "properties": {
                  "a": {
                    "description": "A number",
                    "type": "number",
                  },
                },
                "required": [
                  "a",
                ],
                "type": "object",
              },
              {
                "description": "B",
                "properties": {
                  "b": {
                    "description": "A string",
                    "type": "string",
                  },
                },
                "required": [
                  "b",
                ],
                "type": "object",
              },
            ],
          },
          "output": {
            "anyOf": [
              {
                "description": "A",
                "properties": {
                  "a": {
                    "description": "A number",
                    "type": "number",
                  },
                },
                "required": [
                  "a",
                ],
                "type": "object",
              },
              {
                "description": "B",
                "properties": {
                  "b": {
                    "description": "A string",
                    "type": "string",
                  },
                },
                "required": [
                  "b",
                ],
                "type": "object",
              },
            ],
          },
          "type": "rpc",
        },
        "echoWithPrefix": {
          "errors": {
            "not": {},
          },
          "init": {
            "properties": {
              "prefix": {
                "type": "string",
              },
            },
            "required": [
              "prefix",
            ],
            "type": "object",
          },
          "input": {
            "properties": {
              "ignore": {
                "type": "boolean",
              },
              "msg": {
                "type": "string",
              },
            },
            "required": [
              "msg",
              "ignore",
            ],
            "type": "object",
          },
          "output": {
            "properties": {
              "response": {
                "type": "string",
              },
            },
            "required": [
              "response",
            ],
            "type": "object",
          },
          "type": "stream",
        },
        "unimplementedSubscription": {
          "errors": {
            "not": {},
          },
          "init": {
            "properties": {},
            "type": "object",
          },
          "output": {
            "properties": {},
            "type": "object",
          },
          "type": "subscription",
        },
        "unimplementedUpload": {
          "errors": {
            "not": {},
          },
          "init": {
            "properties": {},
            "type": "object",
          },
          "input": {
            "properties": {},
            "type": "object",
          },
          "output": {
            "properties": {},
            "type": "object",
          },
          "type": "upload",
        },
      },
    },
  },
}
`;

exports[`serialize service to jsonschema > serialize backwards compatible with v1 1`] = `
{
  "procedures": {
    "add": {
      "errors": {
        "not": {},
      },
      "input": {
        "properties": {
          "n": {
            "type": "number",
          },
        },
        "required": [
          "n",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "result": {
            "type": "number",
          },
        },
        "required": [
          "result",
        ],
        "type": "object",
      },
      "type": "rpc",
    },
    "array": {
      "errors": {
        "not": {},
      },
      "input": {
        "properties": {
          "n": {
            "type": "number",
          },
        },
        "required": [
          "n",
        ],
        "type": "object",
      },
      "output": {
        "items": {
          "type": "number",
        },
        "type": "array",
      },
      "type": "rpc",
    },
    "arrayStream": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "input": {
        "properties": {
          "n": {
            "type": "number",
          },
        },
        "required": [
          "n",
        ],
        "type": "object",
      },
      "output": {
        "items": {
          "type": "number",
        },
        "type": "array",
      },
      "type": "stream",
    },
    "echo": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "input": {
        "properties": {
          "ignore": {
            "type": "boolean",
          },
          "msg": {
            "type": "string",
          },
        },
        "required": [
          "msg",
          "ignore",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "response": {
            "type": "string",
          },
        },
        "required": [
          "response",
        ],
        "type": "object",
      },
      "type": "stream",
    },
    "echoUnion": {
      "description": "Echos back whatever we sent",
      "errors": {
        "not": {},
      },
      "input": {
        "anyOf": [
          {
            "description": "A",
            "properties": {
              "a": {
                "description": "A number",
                "type": "number",
              },
            },
            "required": [
              "a",
            ],
            "type": "object",
          },
          {
            "description": "B",
            "properties": {
              "b": {
                "description": "A string",
                "type": "string",
              },
            },
            "required": [
              "b",
            ],
            "type": "object",
          },
        ],
      },
      "output": {
        "anyOf": [
          {
            "description": "A",
            "properties": {
              "a": {
                "description": "A number",
                "type": "number",
              },
            },
            "required": [
              "a",
            ],
            "type": "object",
          },
          {
            "description": "B",
            "properties": {
              "b": {
                "description": "A string",
                "type": "string",
              },
            },
            "required": [
              "b",
            ],
            "type": "object",
          },
        ],
      },
      "type": "rpc",
    },
    "echoWithPrefix": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {
          "prefix": {
            "type": "string",
          },
        },
        "required": [
          "prefix",
        ],
        "type": "object",
      },
      "input": {
        "properties": {
          "ignore": {
            "type": "boolean",
          },
          "msg": {
            "type": "string",
          },
        },
        "required": [
          "msg",
          "ignore",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "response": {
            "type": "string",
          },
        },
        "required": [
          "response",
        ],
        "type": "object",
      },
      "type": "stream",
    },
    "unimplementedSubscription": {
      "errors": {
        "not": {},
      },
      "input": {
        "properties": {},
        "type": "object",
      },
      "output": {
        "properties": {},
        "type": "object",
      },
      "type": "subscription",
    },
    "unimplementedUpload": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "input": {
        "properties": {},
        "type": "object",
      },
      "output": {
        "properties": {},
        "type": "object",
      },
      "type": "upload",
    },
  },
}
`;

exports[`serialize service to jsonschema > serialize basic service 1`] = `
{
  "procedures": {
    "add": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {
          "n": {
            "type": "number",
          },
        },
        "required": [
          "n",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "result": {
            "type": "number",
          },
        },
        "required": [
          "result",
        ],
        "type": "object",
      },
      "type": "rpc",
    },
    "array": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {
          "n": {
            "type": "number",
          },
        },
        "required": [
          "n",
        ],
        "type": "object",
      },
      "output": {
        "items": {
          "type": "number",
        },
        "type": "array",
      },
      "type": "rpc",
    },
    "arrayStream": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "input": {
        "properties": {
          "n": {
            "type": "number",
          },
        },
        "required": [
          "n",
        ],
        "type": "object",
      },
      "output": {
        "items": {
          "type": "number",
        },
        "type": "array",
      },
      "type": "stream",
    },
    "echo": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "input": {
        "properties": {
          "ignore": {
            "type": "boolean",
          },
          "msg": {
            "type": "string",
          },
        },
        "required": [
          "msg",
          "ignore",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "response": {
            "type": "string",
          },
        },
        "required": [
          "response",
        ],
        "type": "object",
      },
      "type": "stream",
    },
    "echoUnion": {
      "description": "Echos back whatever we sent",
      "errors": {
        "not": {},
      },
      "init": {
        "anyOf": [
          {
            "description": "A",
            "properties": {
              "a": {
                "description": "A number",
                "type": "number",
              },
            },
            "required": [
              "a",
            ],
            "type": "object",
          },
          {
            "description": "B",
            "properties": {
              "b": {
                "description": "A string",
                "type": "string",
              },
            },
            "required": [
              "b",
            ],
            "type": "object",
          },
        ],
      },
      "output": {
        "anyOf": [
          {
            "description": "A",
            "properties": {
              "a": {
                "description": "A number",
                "type": "number",
              },
            },
            "required": [
              "a",
            ],
            "type": "object",
          },
          {
            "description": "B",
            "properties": {
              "b": {
                "description": "A string",
                "type": "string",
              },
            },
            "required": [
              "b",
            ],
            "type": "object",
          },
        ],
      },
      "type": "rpc",
    },
    "echoWithPrefix": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {
          "prefix": {
            "type": "string",
          },
        },
        "required": [
          "prefix",
        ],
        "type": "object",
      },
      "input": {
        "properties": {
          "ignore": {
            "type": "boolean",
          },
          "msg": {
            "type": "string",
          },
        },
        "required": [
          "msg",
          "ignore",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "response": {
            "type": "string",
          },
        },
        "required": [
          "response",
        ],
        "type": "object",
      },
      "type": "stream",
    },
    "unimplementedSubscription": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "output": {
        "properties": {},
        "type": "object",
      },
      "type": "subscription",
    },
    "unimplementedUpload": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "input": {
        "properties": {},
        "type": "object",
      },
      "output": {
        "properties": {},
        "type": "object",
      },
      "type": "upload",
    },
  },
}
`;

exports[`serialize service to jsonschema > serialize service with binary 1`] = `
{
  "procedures": {
    "getFile": {
      "errors": {
        "not": {},
      },
      "init": {
        "properties": {
          "file": {
            "type": "string",
          },
        },
        "required": [
          "file",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "contents": {
            "type": "Uint8Array",
          },
        },
        "required": [
          "contents",
        ],
        "type": "object",
      },
      "type": "rpc",
    },
  },
}
`;

exports[`serialize service to jsonschema > serialize service with errors 1`] = `
{
  "procedures": {
    "divide": {
      "errors": {
        "anyOf": [
          {
            "properties": {
              "code": {
                "const": "DIV_BY_ZERO",
                "type": "string",
              },
              "extras": {
                "properties": {
                  "test": {
                    "type": "string",
                  },
                },
                "required": [
                  "test",
                ],
                "type": "object",
              },
              "message": {
                "type": "string",
              },
            },
            "required": [
              "code",
              "message",
              "extras",
            ],
            "type": "object",
          },
          {
            "properties": {
              "code": {
                "const": "INFINITY",
                "type": "string",
              },
              "message": {
                "type": "string",
              },
            },
            "required": [
              "code",
              "message",
            ],
            "type": "object",
          },
        ],
      },
      "init": {
        "properties": {
          "a": {
            "type": "number",
          },
          "b": {
            "type": "number",
          },
        },
        "required": [
          "a",
          "b",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "result": {
            "type": "number",
          },
        },
        "required": [
          "result",
        ],
        "type": "object",
      },
      "type": "rpc",
    },
    "echo": {
      "errors": {
        "properties": {
          "code": {
            "const": "STREAM_ERROR",
            "type": "string",
          },
          "message": {
            "type": "string",
          },
        },
        "required": [
          "code",
          "message",
        ],
        "type": "object",
      },
      "init": {
        "properties": {},
        "type": "object",
      },
      "input": {
        "properties": {
          "msg": {
            "type": "string",
          },
          "throwError": {
            "type": "boolean",
          },
          "throwResult": {
            "type": "boolean",
          },
        },
        "required": [
          "msg",
          "throwResult",
          "throwError",
        ],
        "type": "object",
      },
      "output": {
        "properties": {
          "response": {
            "type": "string",
          },
        },
        "required": [
          "response",
        ],
        "type": "object",
      },
      "type": "stream",
    },
  },
}
`;
